pkg/registry/core/service/strategy.go

```diff
@@ -315,3 +316,33 @@ func clearClusterIPRelatedFields(newService, oldService *api.Service) {
 		}
 	}
 }
+
+// this func allows user to downgrade a service by just changing
+// IPFamilyPolicy to SingleStack
+func trimFieldsForDualStackDowngrade(newService, oldService *api.Service) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
+		return
+	}
+
+	// not an update
+	if oldService == nil {
+		return
+	}
+
+	oldIsDualStack := oldService.Spec.IPFamilyPolicy != nil &&
+		(*oldService.Spec.IPFamilyPolicy == api.IPFamilyPolicyRequireDualStack ||
+			*oldService.Spec.IPFamilyPolicy == api.IPFamilyPolicyPreferDualStack)
+
+	newIsNotDualStack := newService.Spec.IPFamilyPolicy != nil && *newService.Spec.IPFamilyPolicy == api.IPFamilyPolicySingleStack
```

Comment:
should this handle == nil as single?

Reply:
this is called after first round of validation and defaulting. if it is an update and user sets IPFamilyPolicy == nil the defaulting logic will reset it to required.